packages/jupyterlab-manager/src/manager.ts

```diff
@@ -111,31 +153,111 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
     }
   }
 
+  _handleKernelStatusChange(args: Kernel.Status) {
+    switch (args) {
+    case 'connected':
+      // TODO: should we clear away any old widgets before restoring?
+      this.restoreWidgets(this._context.model);
+      break;
+    case 'restarting':
+      this.disconnect();
+      break;
+    default:
+    }
+  }
+
+  /**
+   * Restore widgets from kernel and saved state.
+   */
+  async restoreWidgets(notebook: INotebookModel): Promise<void> {
+    await this._loadFromKernel();
+    await this._loadFromNotebook(notebook);
+    this._restored.emit();
+  }
+
+  async _loadFromKernel(): Promise<void> {
+    if (!this.context.session.kernel) {
+      return;
+    }
+    const comm_ids = await this._get_comm_info();
+
+    // For each comm id, create the comm, get the state, and create a widget.
+    const widgets_info = await Promise.all(Object.keys(comm_ids).map(async (comm_id) => {
+      const comm = await this._create_comm(this.comm_target_name, comm_id);
```

Comment:
Note: This changes the sync logic (does not await all `_create_comm` first), but I do not think this is significant.

Reply:
right, instead of doing things in batches, it does each as soon as it can. I think it's more streamlined. Note that I did put in a sync barrier so that hydrating widgets works well even when there are links between them.